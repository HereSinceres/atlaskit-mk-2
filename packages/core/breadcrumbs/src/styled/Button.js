// @flow
import * as React from 'react';
import Button from '@atlaskit/button';

// $FlowFixMe
export default React.forwardRef((props, ref) => (
  <Button
    {...props}
    ref={ref}
    theme={(currentTheme, themeProps) => {
      const { buttonStyles, ...rest } = currentTheme(themeProps);
      return {
        buttonStyles: {
          ...buttonStyles,
          ...(props.truncationWidth
            ? { maxWidth: `${props.truncationWidth}px !important` }
            : { flexShrink: 1, minWidth:0 }),
        },
        ...rest,
      }
    }}
  />
));
