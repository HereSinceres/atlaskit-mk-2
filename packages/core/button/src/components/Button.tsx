import * as React from 'react';
import { css } from 'emotion';
import {
  withAnalyticsEvents,
  withAnalyticsContext,
  createAndFireEvent,
} from '@atlaskit/analytics-next';
import {
  name as packageName,
  version as packageVersion,
} from '../../package.json';
import GlobalTheme from '@atlaskit/theme';
import { Theme } from '../theme';
import { mapAttributesToState, filterProps } from './utils';
import {
  ButtonContent,
  ButtonWrapper,
  IconWrapper,
  LoadingSpinner,
} from '../styled';
import { withDefaultProps } from '@atlaskit/type-helpers';
import { ButtonProps, ThemeMode } from '../types';

export type ButtonState = {
  isHover: boolean;
  isActive: boolean;
  isFocus: boolean;
};

export const defaultProps: Pick<
  ButtonProps,
  | 'appearance'
  | 'isDisabled'
  | 'isSelected'
  | 'isLoading'
  | 'spacing'
  | 'type'
  | 'shouldFitContainer'
  | 'autoFocus'
> = {
  appearance: 'default',
  isDisabled: false,
  isSelected: false,
  isLoading: false,
  spacing: 'default',
  type: 'button',
  shouldFitContainer: false,
  autoFocus: false,
};

export class Button extends React.Component<ButtonProps, ButtonState> {
  state = {
    isActive: false,
    isFocus: false,
    isHover: false,
  };

  onMouseEnter = () => {
    this.setState({ isHover: true });
  };

  onMouseLeave: React.MouseEventHandler = () => {
    this.setState({ isHover: false, isActive: false });
  };

  onMouseDown: React.MouseEventHandler = event => {
    event.preventDefault();
    this.setState({ isActive: true });
  };

  onMouseUp = () => this.setState({ isActive: false });

  onFocus: React.FocusEventHandler<HTMLButtonElement> = event => {
    this.setState({ isFocus: true });
    if (this.props.onFocus) {
      this.props.onFocus(event);
    }
  };

  onBlur: React.FocusEventHandler<HTMLButtonElement> = event => {
    this.setState({ isFocus: false });
    if (this.props.onBlur) {
      this.props.onBlur(event);
    }
  };

  isInteractive = () => !this.props.isDisabled && !this.props.isLoading;

  // Swallow click events when the button is disabled
  // to prevent inner child clicks bubbling up.
  onInnerClick: React.MouseEventHandler<HTMLButtonElement> = e => {
    if (!this.isInteractive()) e.stopPropagation();
    return true;
  };

  render() {
    const {
      appearance,
      children,
      component,
      href,
      iconAfter,
      iconBefore,
      isDisabled,
      isLoading,
      isSelected,
      shouldFitContainer,
      spacing,
      theme,
    } = this.props;
    const attributes = { ...this.state, isSelected, isDisabled };

    const iconIsOnlyChild: boolean = !!(
      (iconBefore && !iconAfter && !children) ||
      (iconAfter && !iconBefore && !children)
    );

    const getElement = () => {
      if (href) {
        return isDisabled ? 'span' : 'a';
      }
      return 'button';
    };

    const StyledButton = component || getElement();

    const specifiers = (styles: {}) => {
      if (StyledButton === 'a') {
        return {
          'a&': styles,
        };
      } else if (StyledButton === component) {
        return {
          '&, a&, &:hover, &:active, &:focus': styles,
        };
      }
      return styles;
    };

    return (
      <Theme.Provider value={theme}>
        <GlobalTheme.Consumer>
          {({ mode }: { mode: ThemeMode }) => (
            <Theme.Consumer
              appearance={appearance}
              mode={mode}
              state={mapAttributesToState(attributes)}
              {...this.props}
            >
              {({ button }) => (
                <StyledButton
                  {...filterProps(this.props, StyledButton)}
                  onMouseEnter={this.onMouseEnter}
                  onMouseLeave={this.onMouseLeave}
                  onMouseDown={this.onMouseDown}
                  onMouseUp={this.onMouseUp}
                  onFocus={this.onFocus}
                  onBlur={this.onBlur}
                  className={css(specifiers(button))}
                >
                  <ButtonWrapper
                    onClick={this.onInnerClick}
                    fit={!!shouldFitContainer}
                  >
                    {isLoading && (
                      <LoadingSpinner
                        spacing={spacing}
                        appearance={appearance}
                        isSelected={isSelected}
                        isDisabled={isDisabled}
                      />
                    )}
                    {iconBefore && (
                      <IconWrapper
                        isLoading={isLoading}
                        spacing={spacing}
                        isOnlyChild={iconIsOnlyChild}
                      >
                        {iconBefore}
                      </IconWrapper>
                    )}
                    {children && (
                      <ButtonContent
                        isLoading={isLoading}
                        followsIcon={!!iconBefore}
                        spacing={spacing}
                      >
                        {children}
                      </ButtonContent>
                    )}
                    {iconAfter && (
                      <IconWrapper
                        isLoading={isLoading}
                        spacing={spacing}
                        isOnlyChild={iconIsOnlyChild}
                      >
                        {iconAfter}
                      </IconWrapper>
                    )}
                  </ButtonWrapper>
                </StyledButton>
              )}
            </Theme.Consumer>
          )}
        </GlobalTheme.Consumer>
      </Theme.Provider>
    );
  }
}

const DefaultedButton = withDefaultProps(defaultProps, Button);
const createAndFireEventOnAtlaskit = createAndFireEvent('atlaskit');

export default withAnalyticsContext({
  componentName: 'button',
  packageName,
  packageVersion,
})(
  withAnalyticsEvents({
    onClick: createAndFireEventOnAtlaskit({
      action: 'clicked',
      actionSubject: 'button',

      attributes: {
        componentName: 'button',
        packageName,
        packageVersion,
      },
    }),
  })(DefaultedButton),
);
