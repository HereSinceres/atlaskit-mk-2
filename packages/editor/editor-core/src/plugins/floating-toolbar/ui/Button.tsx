import * as React from 'react';
import { ReactElement, MouseEvent } from 'react';

import Tooltip from '@atlaskit/tooltip';
import UiButton from '@atlaskit/button';
import { colors, themed } from '@atlaskit/theme';

import styled from 'styled-components';
import { hexToRgba } from '@atlaskit/editor-common';

const editorButtonTheme = {
  background: {
    danger: {
      default: themed({ light: 'inherit', dark: 'inherit' }),
      hover: themed({ light: colors.N30A, dark: colors.N30A }),
      active: themed({
        light: hexToRgba(colors.B75, 0.6),
        dark: hexToRgba(colors.B75, 0.6),
      }),
    },
  },
  color: {
    danger: {
      hover: themed({ light: colors.R300, dark: colors.R300 }),
      active: themed({ light: colors.R300, dark: colors.R300 }),
    },
  },
};

const Button = styled(UiButton)`
  padding: 0 2px;

  &[href] {
    padding: 0 2px;
  }
`;

const getStyle = (styles, appearance, state) => {
  if (!styles[appearance] || !styles[appearance][state]) return 'initial';
  return styles[appearance][state];
};

export type ButtonAppearance = 'subtle' | 'danger';
export interface Props {
  title?: string;
  icon?: ReactElement<any>;
  iconAfter?: ReactElement<any>;
  onClick: React.MouseEventHandler;
  onMouseEnter?: <T>(event: MouseEvent<T>) => void;
  onMouseLeave?: <T>(event: MouseEvent<T>) => void;
  selected?: boolean;
  disabled?: boolean;
  appearance?: ButtonAppearance;
  href?: string;
  target?: string;
  children?: React.ReactNode;
}

export default ({
  title,
  icon,
  iconAfter,
  onClick,
  onMouseEnter,
  onMouseLeave,
  selected,
  disabled,
  href,
  target,
  appearance = 'subtle',
  children,
}: Props) => {
  return (
    <Tooltip content={title} hideTooltipOnClick={true} position="top">
      <div onMouseEnter={onMouseEnter} onMouseLeave={onMouseLeave}>
        <Button
          theme={(adgTheme, { appearance = 'default', state = 'default' }) => {
            const { buttonStyles: adgButtonStyles } = adgTheme({
              appearance,
              state,
            });

            return {
              buttonStyles: {
                ...adgButtonStyles,
                ...{
                  background: getStyle(
                    editorButtonTheme.background,
                    appearance,
                    state,
                  ),
                  color: getStyle(editorButtonTheme.color, appearance, state),
                },
              },
              iconStyles: {},
              spinnerStyles: {},
            };
          }}
          aria-label={title}
          spacing="compact"
          href={href}
          target={target}
          appearance={appearance}
          aria-haspopup={true}
          iconBefore={icon}
          iconAfter={iconAfter}
          onClick={onClick}
          isSelected={selected}
          isDisabled={disabled}
        >
          {children}
        </Button>
      </div>
    </Tooltip>
  );
};
