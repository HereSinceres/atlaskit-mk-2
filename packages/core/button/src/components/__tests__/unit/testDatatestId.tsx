import { mount } from 'enzyme';
// @eslint
import { render, waitForElement } from '@testing-library/react';
import React from 'react';
import cases from 'jest-in-case';
import Button from '../../Button';

test('It should not generate data-testid', () => {
  const wrapper = mount(<Button>Button</Button>);
  expect(wrapper).toBeDefined();
  expect(wrapper.text()).toBe('Button');
  expect(wrapper.prop('data-testid')).toBeUndefined();
});
// TODO: To investigate why using the render from react-testing-library does not found the `data-testid`.
describe.skip('Button should be found by data-testid', () => {
  test('Using getByTestId()', async () => {
    const { getByTestId } = render(
      <Button data-testid="iamTheDataTestId">Button</Button>,
    );
    await waitForElement(() =>
      expect(getByTestId('iamTheDataTestId')).toBeTruthy(),
    );
  });
  test('Using container and getByTestId()', () => {
    const { getByTestId, container } = render(
      <Button data-testid="iamTheDataTestId">Button</Button>,
    );
    // @ts-ignore
    expect(getByTestId(container, 'iamTheDataTestId')).toBe('Button');
  });
  test('Using container snapshot', () => {
    const { container } = render(
      <Button data-testid="iamTheDataTestId">Button</Button>,
    );
    expect(container).toMatchSnapshot();
  });
});

test('Button snapshot should be same with data-testid ', () => {
  const wrapper = mount(<Button data-testid="iamTheDataTestId">Button</Button>);
  expect(wrapper).toMatchSnapshot();
});

describe('Button with different data-testid', () => {
  cases(
    'should be generated',
    ({ key }: { key: string }) => {
      const wrapper = mount(<Button data-testid={key}>Button</Button>);
      expect(wrapper.prop('data-testid')).toBeDefined();
    },
    [
      { key: 'josefGiTan' },
      { key: 'ZZZZŹŽ;;;;' },
      { key: '@3$&&&&Helooo' },
      { key: '126^^^' },
      { key: 123 },
    ],
  );
});
