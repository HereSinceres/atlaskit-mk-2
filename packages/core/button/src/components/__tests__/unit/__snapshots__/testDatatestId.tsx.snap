// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button snapshot should be same with data-testid  1`] = `
.emotion-2 {
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  border-width: 0;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: inherit;
  font-style: normal;
  font-weight: normal;
  max-width: 100%;
  outline: none !important;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
  background: rgba(9,30,66,0.04);
  border-radius: 3px;
  box-shadow: 0 0 0 2px inherit;
  color: #505F79 !important;
  cursor: default;
  height: 2.2857142857142856em;
  line-height: 2.2857142857142856em;
  padding: 0 8px;
  -webkit-transition: background 0.1s ease-out,box-shadow 0.15s cubic-bezier(0.47,0.03,0.49,1.38);
  transition: background 0.1s ease-out,box-shadow 0.15s cubic-bezier(0.47,0.03,0.49,1.38);
  -webkit-transition-duration: 0.1s,0.15s;
  transition-duration: 0.1s,0.15s;
  vertical-align: middle;
  width: auto;
}

.emotion-2::-moz-focus-inner {
  border: 0;
  margin: 0;
  padding: 0;
}

.emotion-2:hover {
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
}

.emotion-1 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  max-width: 100%;
  position: relative;
}

.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  margin: 0 4px;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  -webkit-transition: opacity 0.3s;
  transition: opacity 0.3s;
  opacity: 1;
}

<WithAnalyticsContext(WithAnalyticsEvents(Button))
  data-testid="iamTheDataTestId"
>
  <AnalyticsContext
    data={
      Object {
        "componentName": "button",
        "packageName": "@atlaskit/button",
        "packageVersion": "999.9.9",
      }
    }
  >
    <WithAnalyticsEvents(Button)
      data-testid="iamTheDataTestId"
    >
      <AnalyticsContextConsumer
        createEventMap={
          Object {
            "onClick": [Function],
          }
        }
        wrappedComponentProps={
          Object {
            "children": "Button",
            "data-testid": "iamTheDataTestId",
          }
        }
      >
        <Button
          createAnalyticsEvent={[Function]}
          data-testid="iamTheDataTestId"
          onClick={[Function]}
        >
          <Button
            appearance="default"
            autoFocus={false}
            consumerRef={null}
            createAnalyticsEvent={[Function]}
            data-testid="iamTheDataTestId"
            isDisabled={false}
            isLoading={false}
            isSelected={false}
            onClick={[Function]}
            shouldFitContainer={false}
            spacing="default"
            theme={[Function]}
            type="button"
          >
            <Provider
              value={[Function]}
            >
              <Consumer>
                <Consumer
                  appearance="default"
                  autoFocus={false}
                  consumerRef={null}
                  createAnalyticsEvent={[Function]}
                  data-testid="iamTheDataTestId"
                  iconIsOnlyChild={false}
                  isDisabled={false}
                  isLoading={false}
                  isSelected={false}
                  mode="light"
                  onClick={[Function]}
                  shouldFitContainer={false}
                  spacing="default"
                  state="default"
                  theme={[Function]}
                  type="button"
                >
                  <button
                    autoFocus={false}
                    className="emotion-2"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onMouseDown={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    type="button"
                  >
                    <Component
                      fit={false}
                      onClick={[Function]}
                    >
                      <span
                        className="emotion-1"
                        onClick={[Function]}
                      >
                        <Component
                          followsIcon={false}
                          isLoading={false}
                          spacing="default"
                        >
                          <span
                            className="emotion-0"
                          >
                            Button
                          </span>
                        </Component>
                      </span>
                    </Component>
                  </button>
                </Consumer>
              </Consumer>
            </Provider>
          </Button>
        </Button>
      </AnalyticsContextConsumer>
    </WithAnalyticsEvents(Button)>
  </AnalyticsContext>
</WithAnalyticsContext(WithAnalyticsEvents(Button))>
`;
